In [1]:
import os
import pandas as pd
import numpy as np
#import pandas_profiling
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoCV
from sklearn import datasets
from collections import Counter

from joblib import dump
from sklearn.externals import joblib

#from keras.models import model_from_json

#Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB

#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import cross_val_score


/home/estudiante/fcrv/jupyter_notebooks/data_analisis_env_py3/lib/python3.5/site-packages/sklearn/externals/joblib/__init__.py:15: FutureWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=FutureWarning)


# Se importa el Dataset con informacion de nubes
Como se ha descrito, las imagenes descargadas de los portales de imagenes satelitales presentan información en multiples bandas. Es con esta información que los algoritmos de clasificación pueden emitir un juicio para determinar a que tipo de información puede corresponder un pixel: bosque, no busque,nubes,agua, agua_bosque,ninguno.
En este escenario se plantea probar el clasificador Naive Bayes con información de imagenes que contengan nubes y agregar una nueva clase: Nubes. 
Este escenario se plantea en respuesta a que la clasificación de imagenes satelitales cuyas nubes fueron removidas resultaban un problema facil de clasificar para este tipo de modelo. 

In [2]:
path = './satellite_dataset-con-nubes.csv'

satellite = pd.read_csv(path)
satellite.head(100)

blue   green     red     nir   swir1   swir2  wofs  bosque  wofs_bosque  \
0    208.0   168.0    88.0    86.0    91.0    72.0     1       0            0   
1    170.0   118.0    40.0    27.0    61.0    53.0     1       0            0   
2    357.0   247.0   143.0   119.0   116.0    90.0     1       0            0   
3    263.0   180.0    87.0    59.0    64.0    45.0     1       0            0   
4   3359.0  3167.0  3127.0  3150.0  2731.0  2703.0     0       0            0   
5    904.0   707.0   549.0   450.0   316.0   222.0     1       0            0   
6    224.0   148.0    61.0    41.0    66.0    54.0     1       0            0   
7    439.0   350.0   214.0   160.0   140.0   104.0     1       0            0   
8   1635.0  2032.0  1949.0  4395.0  3198.0  2215.0     0       0            0   
9   5840.0  6109.0  6246.0  7067.0  5533.0  4197.0     0       0            0   
10  7555.0  7576.0  7624.0  7883.0  4945.0  3243.0     0       0            0   
11   176.0   409.0   228.0  3247.0  1310.0   508.0     0       1            0   
12   205.0   149.0    68.0    59.0    83.0    64.0     1       0            0   
13   511.0   549.0   371.0  1945.0   771.0   302.0     0       1            0   
14  4578.0  4743.0  4827.0  5900.0  5068.0  4050.0     0       0            0   
15   741.0   582.0   430.0   356.0   249.0   190.0     1       0            0   
16  5647.0  5881.0  6036.0  7161.0  6192.0  5176.0     0       0            0   
17   168.0   121.0    32.0    18.0    52.0    50.0     1       0            0   
18    92.0   334.0   238.0  2321.0  1156.0   556.0     0       1            0   
19  1564.0  1598.0  1415.0  3343.0  2155.0  1443.0     0       0            0   
20   554.0   617.0   447.0  2060.0   969.0   531.0     0       0            0   
21  2933.0  2890.0  2842.0  4381.0  3534.0  2769.0     0       0            0   
22   570.0   566.0   391.0  1931.0   817.0   343.0     0       0            0   
23   268.0   443.0   248.0  3544.0  1661.0   619.0     0       1            0   
24   495.0   355.0   229.0   173.0   158.0   112.0     1       0            0   
25   264.0   178.0    91.0    73.0    94.0    76.0     1       0            0   
26  4377.0  4447.0  4463.0  5682.0  4690.0  4229.0     0       0            0   
27  6451.0  6734.0  6936.0  7740.0  6173.0  4572.0     0       0            0   
28   588.0   672.0   553.0  1779.0   992.0   586.0     0       0            0   
29  6535.0  6935.0  7112.0  7732.0  4443.0  3028.0     0       0            0   
..     ...     ...     ...     ...     ...     ...   ...     ...          ...   
70  3154.0  3284.0  3231.0  4752.0  3835.0  2880.0     0       0            0   
71   257.0   427.0   251.0  3429.0  1617.0   616.0     0       1            0   
72  5824.0  5999.0  6090.0  6856.0  5404.0  3950.0     0       0            0   
73  4454.0  4556.0  4503.0  5522.0  4685.0  3961.0     0       0            0   
74   492.0   474.0   305.0  1572.0   558.0   223.0     0       0            0   
75   303.0   478.0   419.0   601.0   379.0   230.0     0       0            1   
76   229.0   165.0    85.0    69.0    90.0    75.0     1       0            0   
77  1226.0  1150.0  1037.0  2023.0  1148.0   758.0     0       0            0   
78  3268.0  3442.0  3430.0  5148.0  3609.0  2541.0     0       0            0   
79  4120.0  4397.0  4474.0  5820.0  5458.0  4636.0     0       0            0   
80   466.0   493.0   294.0  2104.0   768.0   290.0     0       1            0   
81   191.0   128.0    45.0    26.0    59.0    50.0     1       0            0   
82   258.0   172.0    88.0    70.0    94.0    75.0     1       0            0   
83   194.0   132.0    44.0    23.0    49.0    38.0     1       0            0   
84   318.0   213.0   117.0    94.0   104.0    77.0     1       0            0   
85  6245.0  6501.0  6635.0  7547.0  6000.0  4423.0     0       0            0   
86   219.0   137.0    57.0    39.0    64.0    50.0     1       0            0   
87   157.0   345.0   201.0  2617.0  1

# Filtro de valores negativos
Se realiza una filtro de las columnas que en alguna banda contengan un valor negativo porque el clasificador multinomial Naive bayes no funciona bien con esta información.

In [3]:
satellite = satellite[satellite["blue"] > 0]
satellite = satellite[satellite["red"] > 0]
satellite = satellite[satellite["green"] > 0]
satellite = satellite[satellite["nir"] > 0]
satellite = satellite[satellite["swir1"] > 0]
satellite = satellite[satellite["swir2"] > 0]

In [4]:
#satellite['wofs'].value_counts()
#satellite['bosque'].value_counts()
#satellite['wofs_bosque'].value_counts()
#satellite['cloud'].value_counts()
#satellite['ninguno'].value_counts()

# Se reemplaza los valores en columnas de clases
Como parte de preparación de los datos para crear el modelo se deben hacer algunos ajustes a la forma en que el dataset muestra la infiromación.Como se puede observar el dataset, las columnas de las clases se encuentran categorizadas por un valor binario. Lo que se realiza a continuación es reemplazar estos valores binarios por NaN en caso de ser "Cero" y el nombre de la clase en caso de ser "Uno".

In [5]:
# Reemplazar el valor de columna Wofs por NaN(0) y Wofs(1)
satellite['wofs'] = satellite['wofs'].replace(0,np.nan)
satellite['wofs'] = satellite['wofs'].replace(1,'wofs')


In [6]:
# Reemplazar el valor de columna bosque por NaN(0) y bosque(1)
satellite['bosque'] = satellite['bosque'].replace(0,np.nan)
satellite['bosque'] = satellite['bosque'].replace(1,'bosque')

In [7]:
# Reemplazar el valor de columna wofs_bosque por NaN(0) y wofs_bosque(1)
satellite['wofs_bosque'] = satellite['wofs_bosque'].replace(0,np.nan)
satellite['wofs_bosque'] = satellite['wofs_bosque'].replace(1,'wofs_bosque')

In [8]:
# Reemplazar el valor de columna ninguno por NaN(0) y ninguno(1)
satellite['ninguno'] = satellite['ninguno'].replace(0,np.nan)
satellite['ninguno'] = satellite['ninguno'].replace(1,'ninguno')

In [9]:
# Reemplazar el valor de columna ninguno por NaN(0) y cloud(1)
satellite['cloud'] = satellite['cloud'].replace(0,np.nan)
satellite['cloud'] = satellite['cloud'].replace(1,'cloud')

# Nueva columna de clasificación
Como ultima parte para la preparación de los datos debemos tener la clasificación en una sola columna. Procedemos a crearla con el nombre "label" y alli asignaremos el valor categorizado de cada conjunto de atributos.

In [10]:
#Se crea una nueva columna para almacenar el valor de etiqueta
satellite['label']=np.nan

#Se indica en la columna 'label', la clasificacion de la columna 'wofs'. Se esribe "Wofs" en columna 'label' si en  
#la columna 'wofs' esta este valor
for label in [col for col in satellite.columns if 'wofs' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'bosque'. Se esribe "bosque" en columna 'label' si en  
#la columna 'bosque' esta este valor
for label in [col for col in satellite.columns if 'bosque' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'wofs_bosque'. Se esribe "wofs_bosque" en columna 'label' si en  
#la columna 'wofs_bosque' esta este valor
for label in [col for col in satellite.columns if 'wofs_bosque' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'ninguno'. Se esribe "ninguno" en columna 'label' si en  
#la columna 'ninguno' esta este valor
for label in [col for col in satellite.columns if 'ninguno' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'cloud'. Se esribe "cloud" en columna 'label' si en  
#la columna 'cloud' esta este valor
for label in [col for col in satellite.columns if 'cloud' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
satellite.head(20)

blue   green     red     nir   swir1   swir2  wofs  bosque wofs_bosque  \
0    208.0   168.0    88.0    86.0    91.0    72.0  wofs     NaN         NaN   
1    170.0   118.0    40.0    27.0    61.0    53.0  wofs     NaN         NaN   
2    357.0   247.0   143.0   119.0   116.0    90.0  wofs     NaN         NaN   
3    263.0   180.0    87.0    59.0    64.0    45.0  wofs     NaN         NaN   
4   3359.0  3167.0  3127.0  3150.0  2731.0  2703.0   NaN     NaN         NaN   
5    904.0   707.0   549.0   450.0   316.0   222.0  wofs     NaN         NaN   
6    224.0   148.0    61.0    41.0    66.0    54.0  wofs     NaN         NaN   
7    439.0   350.0   214.0   160.0   140.0   104.0  wofs     NaN         NaN   
8   1635.0  2032.0  1949.0  4395.0  3198.0  2215.0   NaN     NaN         NaN   
9   5840.0  6109.0  6246.0  7067.0  5533.0  4197.0   NaN     NaN         NaN   
10  7555.0  7576.0  7624.0  7883.0  4945.0  3243.0   NaN     NaN         NaN   
11   176.0   409.0   228.0  3247.0  1310.0   508.0   NaN  bosque         NaN   
12   205.0   149.0    68.0    59.0    83.0    64.0  wofs     NaN         NaN   
13   511.0   549.0   371.0  1945.0   771.0   302.0   NaN  bosque         NaN   
14  4578.0  4743.0  4827.0  5900.0  5068.0  4050.0   NaN     NaN         NaN   
15   741.0   582.0   430.0   356.0   249.0   190.0  wofs     NaN         NaN   
16  5647.0  5881.0  6036.0  7161.0  6192.0  5176.0   NaN     NaN         NaN   
17   168.0   121.0    32.0    18.0    52.0    50.0  wofs     NaN         NaN   
18    92.0   334.0   238.0  2321.0  1156.0   556.0   NaN  bosque         NaN   
19  1564.0  1598.0  1415.0  3343.0  2155.0  1443.0   NaN     NaN         NaN   

    ninguno  cloud    label  
0       NaN    NaN     wofs  
1       NaN    NaN     wofs  
2       NaN    NaN     wofs  
3       NaN    NaN     wofs  
4       NaN  cloud    cloud  
5       NaN    NaN     wofs  
6       NaN    NaN     wofs  
7       NaN    NaN     wofs  
8   ninguno    NaN  ninguno  
9       NaN  cloud    cloud  
10      NaN  cloud    cloud  
11      NaN    NaN   bosque  
12      NaN    NaN     wofs  
13      NaN    NaN   bosque  
14      NaN  cloud    cloud  
15      NaN    NaN     wofs  
16      NaN  cloud    cloud  
17      NaN    NaN     wofs  
18      NaN    NaN   bosque  
19      NaN  cloud    cloud

wofs
cloud
ninguno
boque
wofs_bosque

In [11]:
# Se crea el conjunto de clasificacion y el de entrenamiento
Y = satellite["label"]
X = satellite.drop(columns=["label","wofs","bosque","wofs_bosque","cloud","ninguno"], axis=1)
Y.value_counts()

cloud          395787
wofs           340403
bosque         215233
ninguno         45327
wofs_bosque       181
Name: label, dtype: int64

# Se crean los conjuntos de entrenamiento y de prueba
Para hacer esto se deben dividir los datos en 4 conjuntos de datos de la siguiente manera: Conjuntos de entrenamiento(X_train, y_train) y Conjunto de test(X_test, y_test)

In [12]:
# Se dividen los datos en los conjuntos de Train y test
X_train, X_test, y_train, y_test = train_test_split(
    X,
    Y,
    test_size=0.3,
    random_state=0
)

In [13]:
X_test

blue   green     red     nir   swir1   swir2
320143   221.0   161.0    79.0    65.0    89.0    76.0
39220   5598.0  5827.0  5909.0  6855.0  5425.0  4121.0
985443  3079.0  3231.0  3193.0  4799.0  3636.0  2970.0
681309  4259.0  4450.0  4402.0  5495.0  4105.0  2919.0
700134   219.0   384.0   239.0  3011.0  1514.0   601.0
514756   196.0   149.0    68.0    54.0    80.0    66.0
944640   212.0   383.0   221.0  3066.0  1278.0   502.0
913436   708.0   671.0   512.0  2502.0  1160.0   550.0
927309  3070.0  3102.0  3082.0  4944.0  3876.0  2966.0
516675   237.0   334.0   174.0  1686.0   642.0   235.0
375193   196.0   166.0    79.0    74.0    86.0    72.0
549948  1398.0  1276.0  1203.0  1207.0  1460.0  1259.0
214648   249.0   182.0    89.0    73.0    95.0    71.0
146169   231.0   148.0    58.0    39.0    61.0    50.0
416196  6756.0  7151.0  7286.0  7823.0  5466.0  4202.0
148581   126.0   210.0    92.0  1423.0   478.0   201.0
874345   200.0   123.0    44.0    29.0    61.0    48.0
840112   153.0   356.0   195.0  2603.0   965.0   356.0
621568   166.0   437.0   235.0  3908.0  1328.0   454.0
989775  1723.0  1555.0  1442.0  1414.0  1388.0   979.0
996506  3810.0  4045.0  3914.0  5298.0  4335.0  3513.0
825551  2672.0  2877.0  2840.0  4689.0  3618.0  2881.0
780541   201.0   132.0    48.0    30.0    59.0    54.0
512451  1763.0  1844.0  1859.0  4125.0  2750.0  1806.0
225029   190.0   148.0    63.0    48.0    62.0    48.0
152186   900.0  1021.0   847.0  3269.0  1722.0   889.0
273858  4799.0  4998.0  5194.0  5646.0  6076.0  4724.0
895864  3777.0  3787.0  3843.0  5253.0  3889.0  3018.0
677169  6863.0  7052.0  7320.0  8185.0  6541.0  4968.0
748018   401.0   327.0   143.0    89.0    80.0    57.0
...        ...     ...     ...     ...     ...     ...
821481   277.0   190.0   101.0    85.0    96.0    74.0
257829  7091.0  7306.0  7483.0  8026.0  5403.0  3745.0
937050  5296.0  5344.0  5360.0  6172.0  4737.0  3283.0
624765   262.0   174.0    89.0    69.0    92.0    72.0
914965   880.0  1031.0   852.0  2798.0  1530.0   906.0
229550   255.0   431.0   230.0  3611.0  1568.0   603.0
942430  3407.0  3435.0  3380.0  4727.0  3780.0  2868.0
815307   284.0   264.0   164.0   149.0   146.0   115.0
91960    361.0   221.0   133.0   125.0   109.0    89.0
709128   159.0   113.0    29.0    13.0    41.0    35.0
311443  4546.0  4540.0  4550.0  5520.0  4702.0  3874.0
772444   386.0   291.0   135.0    92.0    87.0    70.0
950020   271.0   434.0   239.0  3167.0  1439.0   557.0
572626   233.0   435.0   250.0  3710.0  1548.0   566.0
42690    181.0   484.0   266.0  3667.0  1507.0   552.0
260699  3835.0  3938.0  3907.0  5613.0  4484.0  3579.0
457352   404.0   399.0   234.0  1764.0   643.0   270.0
503600   224.0   153.0    61.0    50.0    67.0    56.0
252506  5665.0  5803.0  5887.0  6678.0  5264.0  3918.0
445252   106.0   419.0   202.0  2830.0  1087.0   387.0
269617  5353.0  5438.0  5495.0  6521.0  5247.0  3914.0
647683   275.0   532.0   357.0  3748.0  1631.0   662.0
605316   244.0   175.0    81.0    51.0    68.0    51.0
641103   121.0   274.0   167.0  1820.0   806.0   290.0
806561  8284.0  8651.0  8964.0  9704.0  6962.0  4772.0
311208   259.0   189.0    88.0    67.0    82.0    60.0
890562  1898.0  1967.0  1876.0  1895.0  2218.0  1868.0
522611  5315.0  5397.0  5475.0  6235.0  5245.0  4278.0
240868   366.0   380.0   253.0   309.0   123.0    68.0
3828     185.0   333.0   206.0  2671.0  1191.0   456.0

[299080 rows x 6 columns]

In [14]:
y_test

320143       wofs
39220       cloud
985443      cloud
681309      cloud
700134     bosque
514756       wofs
944640     bosque
913436    ninguno
927309      cloud
516675     bosque
375193       wofs
549948      cloud
214648       wofs
146169       wofs
416196      cloud
148581     bosque
874345       wofs
840112     bosque
621568     bosque
989775      cloud
996506      cloud
825551      cloud
780541       wofs
512451      cloud
225029       wofs
152186     bosque
273858      cloud
895864      cloud
677169      cloud
748018       wofs
           ...   
821481       wofs
257829      cloud
937050      cloud
624765       wofs
914965    ninguno
229550     bosque
942430      cloud
815307       wofs
91960        wofs
709128       wofs
311443      cloud
772444       wofs
950020     bosque
572626     bosque
42690      bosque
260699      cloud
457352     bosque
503600       wofs
252506      cloud
445252     bosque
269617      cloud
647683     bosque
605316       wofs
641103     bosque
806561    

# Naive Bayes para multiples capas
La clasificación multicapa del algoritmo Naive Bayes es conocida como "Clasificación Naive Bayes Multinomial". Como se puede apreciar la clasificación de nuestro conjunto de caracteristicas puede tener 4 valores: wofs, bosque, wofs_bosque y ninguno. 
En esta parte de creación del modelo se evaluaran dos variaciones del clasificador Naive Bayes: Gausiano y multinomial. En el analisis de resultados se contrastaran los resultados obtenidos para cada uno de ellos.

### Multinomial

In [15]:
#Create a Gaussian Classifier
gnb = MultinomialNB()

In [16]:
#Train the model using the training sets
gnb.fit(X_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [17]:
#Predict the response for test dataset
y_pred = gnb.predict(X_test)

In [18]:
Counter(y_pred)

Counter({'bosque': 60048,
         'cloud': 109882,
         'ninguno': 27271,
         'wofs': 99543,
         'wofs_bosque': 2336})

In [19]:
#wofs
#cloud
#ninguno
#bosque
#wofs_bosque
confusion_matrix(y_test, y_pred)

array([[ 59236,      2,   5475,      2,     10],
       [     7, 105994,  10958,    105,   1701],
       [   805,   1606,  10799,    102,    372],
       [     0,   2267,     24,  99333,    226],
       [     0,     13,     15,      1,     27]])

In [20]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

      bosque       0.99      0.92      0.95     64725
       cloud       0.96      0.89      0.93    118765
     ninguno       0.40      0.79      0.53     13684
        wofs       1.00      0.98      0.99    101850
 wofs_bosque       0.01      0.48      0.02        56

    accuracy                           0.92    299080
   macro avg       0.67      0.81      0.68    299080
weighted avg       0.95      0.92      0.93    299080



In [21]:
# Model Accuracy
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9207870803798315


In [23]:
joblib.dump(gnb, 'model_nb_multinomial.joblib') 

['model_nb_multinomial.joblib']

### Gaussiano

In [15]:
#Create a Gaussian Classifier
gnb = GaussianNB()

In [16]:
#Train the model using the training sets
gnb.fit(X_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [17]:
#Predict the response for test dataset
y_pred = gnb.predict(X_test)

In [18]:
Counter(y_pred)

Counter({'bosque': 60048,
         'cloud': 109882,
         'ninguno': 27271,
         'wofs': 99543,
         'wofs_bosque': 2336})

In [19]:
#wofs
#cloud
#ninguno
#bosque
#wofs_bosque
confusion_matrix(y_test, y_pred)

array([[ 59236,      2,   5475,      2,     10],
       [     7, 105994,  10958,    105,   1701],
       [   805,   1606,  10799,    102,    372],
       [     0,   2267,     24,  99333,    226],
       [     0,     13,     15,      1,     27]])

In [20]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

      bosque       0.99      0.92      0.95     64725
       cloud       0.96      0.89      0.93    118765
     ninguno       0.40      0.79      0.53     13684
        wofs       1.00      0.98      0.99    101850
 wofs_bosque       0.01      0.48      0.02        56

    accuracy                           0.92    299080
   macro avg       0.67      0.81      0.68    299080
weighted avg       0.95      0.92      0.93    299080



In [21]:
# Model Accuracy
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9207870803798315


In [23]:
joblib.dump(gnb, 'model_nb_gaussiano.joblib') 

['model_nb_multinomial.joblib']

## Conclusiones

El rendmiento obtenido con el modelo Multinomial llega a ser del 79.73% y el rendimiento con el modelo Gausiano alcanza el 87.11%. Por lo que logramos determinar que en este escenario funciona mejor el modelo Gassiano que el Multinomial por aproimadamente 8 puntos de diferencia. Revisando la matriz de confusión logramos evidenciar que el modelo Multinomial confunde mucho de los elementos de la clase 3(Ninguno) siendo esto logico porque los valores de caracteristicas de entrada


**Multinomial**

Accuracy: 0.7973101132317454

array([[ 28721,      1,    320,      2,    551],
       [    43, 120350,  39851,    240,   1295],
       [  1895,   8902,   4897,   1003,   1646],
       [     0,   3826,    175,  82221,    278],
       [     0,     12,      7,      9,     50]], dtype=int64)
       
**Gaussiano**

Accuracy: 0.8711520612902681

array([[ 27133,     74,   1321,    349,    718],
       [    68, 134215,  27495,      1,      0],
       [  2570,   1176,  13427,   1065,    105],
       [    95,      0,   2685,  83335,    385],
       [    24,      0,      6,     40,      8]], dtype=int64)

# Referencias
https://www.datacamp.com/community/tutorials/naive-bayes-scikit-learn

https://datascience.stackexchange.com/questions/40345/how-to-convert-multiple-columns-into-single-columns-in-pandas

https://stackoverflow.com/questions/13295735/how-can-i-replace-all-the-nan-values-with-zeros-in-a-column-of-a-pandas-datafra